scripts/Stretch.py

```diff
@@ -0,0 +1,419 @@
+    """
+    Execution part of the stretch algorithm
+    """
+    def __init__(self, line_width, stretch):
+        self.line_width = line_width
+        self.stretch = stretch
+        self.outpos = GCodeStep(0)
+        self.vd1 = np.empty((0, 2)) # Start points of segments
+                                    # of already deposited material for current layer
+        self.vd2 = np.empty((0, 2)) # End points of segments
+                                    # of already deposited material for current layer
+
+    def execute(self, data):
+        """
+        Computes the new X and Y coordinates of all g-code steps
+        """
+        Logger.log("d", "Post stretch with line width=" + str(self.line_width)
+                   + "mm and stretch=" + str(self.stretch)+ "mm")
+        retdata = []
+        layer_steps = []
+        current = GCodeStep(0)
+        layer_z = 0.
+        for layer in data:
+            lines = layer.rstrip("\n").split("\n")
+            for line in lines:
+                current.comment = ""
+                if line.find(";") >= 0:
+                    current.comment = line[line.find(";"):]
+                if _getValue(line, "G") == 0:
+                    current.readStep(line)
+                    onestep = GCodeStep(0)
+                    onestep.copyPosFrom(current)
+                elif _getValue(line, "G") == 1:
+                    current.readStep(line)
+                    onestep = GCodeStep(1)
+                    onestep.copyPosFrom(current)
+                elif _getValue(line, "G") == 92:
+                    current.readStep(line)
+                    onestep = GCodeStep(-1)
+                    onestep.copyPosFrom(current)
+                else:
+                    onestep = GCodeStep(-1)
+                    onestep.copyPosFrom(current)
+                    onestep.comment = line
+                if current.step_z != layer_z:
+                    Logger.log("d", "Layer Z " + "{:.3f}".format(layer_z)
+                               + " " + str(len(layer_steps)) + " steps")
+                    if len(layer_steps):
+                        retdata.append(self.processLayer(layer_steps))
+                    layer_steps = []
+                    layer_z = current.step_z
+                layer_steps.append(onestep)
+        if len(layer_steps):
+            retdata.append(self.processLayer(layer_steps))
+        retdata.append(";Stretch distance " + str(self.stretch) + "\n")
+        return retdata
+
+    def processLayer(self, layer_steps):
+        """
+        Computes the new coordinates of g-code steps
+        for one layer (all the steps at the same Z coordinate)
+        """
+        layergcode = ""
+        self.vd1 = np.empty((0, 2))
+        self.vd2 = np.empty((0, 2))
+        current_e = layer_steps[0].step_e
+        orig_seq = np.empty((0, 2))
+        iflush = 0
+        for i, step in enumerate(layer_steps):
+            if i == 0:
+                current_e = step.step_e
```

Comment:
I suppose that layer switching occurs only during a travel move. Knowing the way curaengine works, it should always be the case *except* in spiralize mode, which I never tried.